// javascript-typescript-langserver/lib/logging.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chalk_1 = require("chalk");
const fs = require("fs");
const util_1 = require("util");
const vscode_languageserver_1 = require("vscode-languageserver");
/**
 * Formats values to a message by pretty-printing objects
 */
function format(values) {
    return values.map(value => (typeof value === 'string' ? value : util_1.inspect(value, { depth: Infinity }))).join(' ');
}
/**
 * A logger implementation that sends window/logMessage notifications to an LSP client
 */
class LSPLogger {
    /**
     * @param client The client to send window/logMessage notifications to
     */
    constructor(client) {
        this.client = client;
    }
    log(...values) {
        try {
            this.client.windowLogMessage({ type: vscode_languageserver_1.MessageType.Log, message: format(values) });
        }
        catch (err) {
            // ignore
        }
    }
    info(...values) {
        try {
            this.client.windowLogMessage({ type: vscode_languageserver_1.MessageType.Info, message: format(values) });
        }
        catch (err) {
            // ignore
        }
    }
    warn(...values) {
        try {
            this.client.windowLogMessage({ type: vscode_languageserver_1.MessageType.Warning, message: format(values) });
        }
        catch (err) {
            // ignore
        }
    }
    error(...values) {
        try {
            this.client.windowLogMessage({ type: vscode_languageserver_1.MessageType.Error, message: format(values) });
        }
        catch (err) {
            // ignore
        }
    }
}
exports.LSPLogger = LSPLogger;
/**
 * Logging implementation that writes to an arbitrary NodeJS stream
 */
class StreamLogger {
    constructor(outStream, errStream) {
        this.outStream = outStream;
        this.errStream = errStream;
    }
    log(...values) {
        try {
            this.outStream.write(chalk_1.default.grey('DEBUG ' + format(values) + '\n'));
        }
        catch (err) {
            // ignore
        }
    }
    info(...values) {
        try {
            this.outStream.write(chalk_1.default.bgCyan('INFO') + '  ' + format(values) + '\n');
        }
        catch (err) {
            // ignore
        }
    }
    warn(...values) {
        try {
            this.errStream.write(chalk_1.default.bgYellow('WARN') + '  ' + format(values) + '\n');
        }
        catch (err) {
            // ignore
        }
    }
    error(...values) {
        try {
            this.errStream.write(chalk_1.default.bgRed('ERROR') + ' ' + format(values) + '\n');
        }
        catch (err) {
            // ignore
        }
    }
}
exports.StreamLogger = StreamLogger;
/**
 * Logger implementation that logs to STDOUT and STDERR depending on level
 */
class StdioLogger extends StreamLogger {
    constructor() {
        super(process.stdout, process.stderr);
    }
}
exports.StdioLogger = StdioLogger;
/**
 * Logger implementation that logs only to STDERR
 */
class StderrLogger extends StreamLogger {
    constructor() {
        super(process.stderr, process.stderr);
    }
}
exports.StderrLogger = StderrLogger;
/**
 * Logger implementation that logs to a file
 */
class FileLogger extends StreamLogger {
    /**
     * @param file Path to the logfile
     */
    constructor(file) {
        const stream = fs.createWriteStream(file);
        super(stream, stream);
    }
}
exports.FileLogger = FileLogger;
/**
 * Logger implementation that wraps another logger and prefixes every message with a given prefix
 */
class PrefixedLogger {
    constructor(logger, prefix) {
        this.logger = logger;
        this.prefix = prefix;
    }
    log(...values) {
        this.logger.log(`[${this.prefix}] ${format(values)}`);
    }
    info(...values) {
        this.logger.info(`[${this.prefix}] ${format(values)}`);
    }
    warn(...values) {
        this.logger.warn(`[${this.prefix}] ${format(values)}`);
    }
    error(...values) {
        this.logger.error(`[${this.prefix}] ${format(values)}`);
    }
}
exports.PrefixedLogger = PrefixedLogger;
/**
 * Logger implementation that does nothing
 */
class NoopLogger {
    log(...values) {
        // empty
    }
    info(...values) {
        // empty
    }
    warn(...values) {
        // empty
    }
    error(...values) {
        // empty
    }
}
exports.NoopLogger = NoopLogger;
