// flow-language-server/lib/pkg/nuclide-flow-rpc/lib/astToOutline.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /**
                                                                                                                                                                                                                                                                   * Copyright (c) 2017-present, Facebook, Inc.
                                                                                                                                                                                                                                                                   * All rights reserved.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * This source code is licensed under the BSD-style license found in the
                                                                                                                                                                                                                                                                   * LICENSE file in the root directory of this source tree. An additional grant
                                                                                                                                                                                                                                                                   * of patent rights can be found in the PATENTS file in the same directory.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * 
                                                                                                                                                                                                                                                                   * @format
                                                                                                                                                                                                                                                                   */

exports.astToOutline = astToOutline;

var _simpleTextBuffer = require('simple-text-buffer');

var _collection = require('nuclide-commons/collection');

var _tokenizedText = require('nuclide-commons/tokenized-text');

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function astToOutline(ast) {
  return {
    outlineTrees: itemsToTrees(ast.body)
  };
}

function itemsToTrees(items) {
  return (0, _collection.arrayCompact)(items.map(itemToTree));
}

function itemToTree(item) {
  if (item == null) {
    return null;
  }
  const extent = getExtent(item);
  switch (item.type) {
    case 'FunctionDeclaration':
    case 'ArrowFunctionExpression':
      return functionOutline(item.id != null ? item.id.name : '', item.params, extent);
    case 'ClassDeclaration':
    case 'ClassExpression':
      const tokenizedText = [(0, _tokenizedText.keyword)('class')];
      let representativeName = undefined;
      if (item.id != null) {
        tokenizedText.push((0, _tokenizedText.whitespace)(' '), (0, _tokenizedText.className)(item.id.name));
        representativeName = item.id.name;
      }
      return _extends({
        kind: 'class',
        tokenizedText,
        representativeName,
        children: itemsToTrees(item.body.body)
      }, extent);
    case 'ClassProperty':
      let paramTokens = [];
      if (item.value && item.value.type === 'ArrowFunctionExpression') {
        paramTokens = [(0, _tokenizedText.plain)('('), ...declarationsTokenizedText(item.value.params), (0, _tokenizedText.plain)(')')];
      }
      return _extends({
        kind: 'property',
        tokenizedText: [(0, _tokenizedText.method)(item.key.name), (0, _tokenizedText.plain)('='), ...paramTokens],
        representativeName: item.key.name,
        children: []
      }, extent);
    case 'MethodDefinition':
      return _extends({
        kind: 'method',
        tokenizedText: [(0, _tokenizedText.method)(item.key.name), (0, _tokenizedText.plain)('('), ...declarationsTokenizedText(item.value.params), (0, _tokenizedText.plain)(')')],
        representativeName: item.key.name,
        children: []
      }, extent);
    case 'ExportDeclaration':
    case 'ExportNamedDeclaration':
      return exportDeclaration(item, extent, Boolean(item.default));
    case 'ExportDefaultDeclaration':
      return exportDeclaration(item, extent, true);
    case 'ExpressionStatement':
      return topLevelExpressionOutline(item);
    case 'TypeAlias':
      return typeAliasOutline(item);
    case 'VariableDeclaration':
      return variableDeclarationOutline(item);
    default:
      return null;
  }
}

function exportDeclaration(item, extent, isDefault) {
  const tree = itemToTree(item.declaration);
  if (tree == null) {
    return null;
  }
  const tokenizedText = [(0, _tokenizedText.keyword)('export'), (0, _tokenizedText.whitespace)(' ')];
  if (isDefault) {
    tokenizedText.push((0, _tokenizedText.keyword)('default'), (0, _tokenizedText.whitespace)(' '));
  }
  // Flow always has tokenizedText
  (0, _assert2.default)(tree.tokenizedText != null);
  tokenizedText.push(...tree.tokenizedText);
  return _extends({
    kind: tree.kind,
    tokenizedText,
    representativeName: tree.representativeName,
    children: tree.children
  }, extent);
}

function declarationReducer(textElements, p, index, declarations) {
  switch (p.type) {
    case 'Identifier':
      textElements.push((0, _tokenizedText.param)(p.name));
      break;
    case 'ObjectPattern':
      textElements.push((0, _tokenizedText.plain)('{'));
      textElements.push(...declarationsTokenizedText(p.properties.map(obj => obj.key)));
      textElements.push((0, _tokenizedText.plain)('}'));
      break;
    case 'ArrayPattern':
      textElements.push((0, _tokenizedText.plain)('['));
      textElements.push(...declarationsTokenizedText(p.elements));
      textElements.push((0, _tokenizedText.plain)(']'));
      break;
    case 'AssignmentPattern':
      return declarationReducer(textElements, p.left, index, declarations);
    case 'RestElement':
      textElements.push((0, _tokenizedText.plain)('...'));
      return declarationReducer(textElements, p.argument, index, declarations);
    default:
      throw new Error(`encountered unexpected argument type ${p.type}`);
  }
  if (index < declarations.length - 1) {
    textElements.push((0, _tokenizedText.plain)(','));
    textElements.push((0, _tokenizedText.whitespace)(' '));
  }
  return textElements;
}

function declarationsTokenizedText(declarations) {
  return declarations.reduce(declarationReducer, []);
}

function getExtent(item) {
  return {
    startPosition: new _simpleTextBuffer.Point(
    // It definitely makes sense that the lines we get are 1-based and the columns are
    // 0-based... convert to 0-based all around.
    item.loc.start.line - 1, item.loc.start.column),
    endPosition: new _simpleTextBuffer.Point(item.loc.end.line - 1, item.loc.end.column)
  };
}

function functionOutline(name, params, extent) {
  return _extends({
    kind: 'function',
    tokenizedText: [(0, _tokenizedText.keyword)('function'), (0, _tokenizedText.whitespace)(' '), (0, _tokenizedText.method)(name), (0, _tokenizedText.plain)('('), ...declarationsTokenizedText(params), (0, _tokenizedText.plain)(')')],
    representativeName: name,
    children: []
  }, extent);
}

function typeAliasOutline(typeAliasExpression) {
  (0, _assert2.default)(typeAliasExpression.type === 'TypeAlias');
  const name = typeAliasExpression.id.name;
  return _extends({
    kind: 'interface',
    tokenizedText: [(0, _tokenizedText.keyword)('type'), (0, _tokenizedText.whitespace)(' '), (0, _tokenizedText.type)(name)],
    representativeName: name,
    children: []
  }, getExtent(typeAliasExpression));
}

function topLevelExpressionOutline(expressionStatement) {
  switch (expressionStatement.expression.type) {
    case 'CallExpression':
      return specOutline(expressionStatement, /* describeOnly */true);
    case 'AssignmentExpression':
      return moduleExportsOutline(expressionStatement.expression);
    default:
      return null;
  }
}

function moduleExportsOutline(assignmentStatement) {
  (0, _assert2.default)(assignmentStatement.type === 'AssignmentExpression');

  const left = assignmentStatement.left;
  if (!isModuleExports(left)) {
    return null;
  }

  const right = assignmentStatement.right;
  if (right.type !== 'ObjectExpression') {
    return null;
  }
  const properties = right.properties;
  return _extends({
    kind: 'module',
    tokenizedText: [(0, _tokenizedText.plain)('module.exports')],
    children: (0, _collection.arrayCompact)(properties.map(moduleExportsPropertyOutline))
  }, getExtent(assignmentStatement));
}

function isModuleExports(left) {
  return left.type === 'MemberExpression' && left.object.type === 'Identifier' && left.object.name === 'module' && left.property.type === 'Identifier' && left.property.name === 'exports';
}

function moduleExportsPropertyOutline(property) {
  (0, _assert2.default)(property.type === 'Property');
  if (property.key.type !== 'Identifier') {
    return null;
  }
  const propName = property.key.name;

  if (property.shorthand) {
    // This happens when the shorthand `{ foo }` is used for `{ foo: foo }`
    return _extends({
      kind: 'method',
      tokenizedText: [(0, _tokenizedText.string)(propName)],
      representativeName: propName,
      children: []
    }, getExtent(property));
  }

  if (property.value.type === 'FunctionExpression' || property.value.type === 'ArrowFunctionExpression') {
    return _extends({
      kind: 'method',
      tokenizedText: [(0, _tokenizedText.method)(propName), (0, _tokenizedText.plain)('('), ...declarationsTokenizedText(property.value.params), (0, _tokenizedText.plain)(')')],
      representativeName: propName,
      children: []
    }, getExtent(property));
  }

  return _extends({
    kind: 'field',
    tokenizedText: [(0, _tokenizedText.string)(propName), (0, _tokenizedText.plain)(':')],
    representativeName: propName,
    children: []
  }, getExtent(property));
}

function specOutline(expressionStatement, describeOnly = false) {
  const expression = expressionStatement.expression;
  if (expression.type !== 'CallExpression') {
    return null;
  }
  const functionName = getFunctionName(expression.callee);
  if (functionName == null) {
    return null;
  }
  if (!isDescribe(functionName)) {
    if (describeOnly || !isIt(functionName)) {
      return null;
    }
  }
  const description = getStringLiteralValue(expression.arguments[0]);
  const specBody = getFunctionBody(expression.arguments[1]);
  if (description == null || specBody == null) {
    return null;
  }
  let children;
  if (isIt(functionName)) {
    children = [];
  } else {
    children = (0, _collection.arrayCompact)(specBody.filter(item => item.type === 'ExpressionStatement').map(item => specOutline(item)));
  }
  return _extends({
    kind: 'function',
    tokenizedText: [(0, _tokenizedText.method)(functionName), (0, _tokenizedText.whitespace)(' '), (0, _tokenizedText.string)(description)],
    representativeName: description,
    children
  }, getExtent(expressionStatement));
}

// Return the function name as written as a string. Intended to stringify patterns like `describe`
// and `describe.only` even though `describe.only` is a MemberExpression rather than an Identifier.
function getFunctionName(callee) {
  switch (callee.type) {
    case 'Identifier':
      return callee.name;
    case 'MemberExpression':
      if (callee.object.type !== 'Identifier' || callee.property.type !== 'Identifier') {
        return null;
      }
      return `${callee.object.name}.${callee.property.name}`;
    default:
      return null;
  }
}

function isDescribe(functionName) {
  switch (functionName) {
    case 'describe':
    case 'fdescribe':
    case 'ddescribe':
    case 'xdescribe':
    case 'describe.only':
    case 'describe.skip':
    case 'test.cb':
    case 'test.serial':
    case 'test.todo':
    case 'test.failing':
    case 'test':
    case 'test.concurrent':
    case 'test.only':
    case 'test.skip':
    case 'suite':
    case 'suite.only':
    case 'suite.skip':
    case 'xtest':
    case 'xtest.concurrent':
    case 'xtest.only':
    case 'xtest.skip':
      return true;
    default:
      return false;
  }
}

function isIt(functionName) {
  switch (functionName) {
    case 'it':
    case 'fit':
    case 'iit':
    case 'pit':
    case 'xit':
    case 'it.only':
    case 'it.skip':
      return true;
    default:
      return false;
  }
}

/** If the given AST Node is a string literal, return its literal value. Otherwise return null */
function getStringLiteralValue(literal) {
  if (literal == null) {
    return null;
  }
  if (literal.type !== 'Literal') {
    return null;
  }
  const value = literal.value;
  if (typeof value !== 'string') {
    return null;
  }
  return value;
}

function getFunctionBody(fn) {
  if (fn == null) {
    return null;
  }
  if (fn.type !== 'ArrowFunctionExpression' && fn.type !== 'FunctionExpression') {
    return null;
  }
  return fn.body.body;
}

function variableDeclarationOutline(declaration) {
  // If there are multiple var declarations in one line, just take the first.
  return variableDeclaratorOutline(declaration.declarations[0], declaration.kind, getExtent(declaration));
}

function variableDeclaratorOutline(declarator, kind, extent) {
  if (declarator.init != null && (declarator.init.type === 'FunctionExpression' || declarator.init.type === 'ArrowFunctionExpression')) {
    return functionOutline(declarator.id.name, declarator.init.params, extent);
  }

  const id = declarator.id;


  const tokenizedText = [(0, _tokenizedText.keyword)(kind), (0, _tokenizedText.whitespace)(' '), ...declarationsTokenizedText([id])];
  const representativeName = id.type === 'Identifier' ? id.name : undefined;
  return _extends({
    kind: kind === 'const' ? 'constant' : 'variable',
    tokenizedText,
    representativeName,
    children: []
  }, extent);
}
