// redux-devtools-inspector/lib/ActionList.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _ReactDOM = require('react/lib/ReactDOM');

var _ReactDOM2 = _interopRequireDefault(_ReactDOM);

var _ActionListRow = require('./ActionListRow');

var _ActionListRow2 = _interopRequireDefault(_ActionListRow);

var _ActionListHeader = require('./ActionListHeader');

var _ActionListHeader2 = _interopRequireDefault(_ActionListHeader);

var _function = require('react-pure-render/function');

var _function2 = _interopRequireDefault(_function);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getTimestamps(actions, actionIds, actionId) {
  var idx = actionIds.indexOf(actionId);
  var prevActionId = actionIds[idx - 1];

  return {
    current: actions[actionId].timestamp,
    previous: idx ? actions[prevActionId].timestamp : 0
  };
}

var ActionList = function (_Component) {
  (0, _inherits3.default)(ActionList, _Component);

  function ActionList() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ActionList);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_Object$getPrototypeO = (0, _getPrototypeOf2.default)(ActionList)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.shouldComponentUpdate = _function2.default, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ActionList, [{
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      if (this.props.lastActionId !== prevProps.lastActionId) {
        this.scrollToBottom();
      }
    }
  }, {
    key: 'scrollToBottom',
    value: function scrollToBottom() {
      var el = _ReactDOM2.default.findDOMNode(this.refs.rows);

      el.scrollTop = el.scrollHeight;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props;
      var styling = _props.styling;
      var actions = _props.actions;
      var actionIds = _props.actionIds;
      var isWideLayout = _props.isWideLayout;
      var onToggleAction = _props.onToggleAction;
      var skippedActionIds = _props.skippedActionIds;
      var selectedActionId = _props.selectedActionId;
      var startActionId = _props.startActionId;
      var _onSelect = _props.onSelect;
      var onSearch = _props.onSearch;
      var searchValue = _props.searchValue;
      var onCommit = _props.onCommit;
      var onSweep = _props.onSweep;

      var lowerSearchValue = searchValue && searchValue.toLowerCase();
      var filteredActionIds = searchValue ? actionIds.filter(function (id) {
        return actions[id].action.type.toLowerCase().indexOf(lowerSearchValue) !== -1;
      }) : actionIds;

      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ key: 'actionList'
        }, styling(['actionList', isWideLayout && 'actionListWide'], isWideLayout)),
        _react2.default.createElement(_ActionListHeader2.default, { styling: styling,
          onSearch: onSearch,
          onCommit: onCommit,
          onSweep: onSweep,
          hasSkippedActions: skippedActionIds.length > 0,
          hasStagedActions: actionIds.length > 1 }),
        _react2.default.createElement(
          'div',
          (0, _extends3.default)({}, styling('actionListRows'), { ref: 'rows' }),
          filteredActionIds.map(function (actionId) {
            return _react2.default.createElement(_ActionListRow2.default, { key: actionId,
              styling: styling,
              isInitAction: !actionId,
              isSelected: startActionId !== null && actionId >= startActionId && actionId <= selectedActionId || actionId === selectedActionId,
              onSelect: function onSelect(e) {
                return _onSelect(e, actionId);
              },
              timestamps: getTimestamps(actions, actionIds, actionId),
              action: actions[actionId].action,
              onViewClick: function onViewClick() {
                return onToggleAction(actionId);
              },
              onCommitClick: function onCommitClick() {
                return onCommit(actionId);
              },
              isSkipped: skippedActionIds.indexOf(actionId) !== -1 });
          })
        )
      );
    }
  }]);
  return ActionList;
}(_react.Component);

exports.default = ActionList;
