// redux-devtools-inspector/lib/ActionListRow.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _function = require('react-pure-render/function');

var _function2 = _interopRequireDefault(_function);

var _dateformat = require('dateformat');

var _dateformat2 = _interopRequireDefault(_dateformat);

var _RightSlider = require('./RightSlider');

var _RightSlider2 = _interopRequireDefault(_RightSlider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BUTTON_SKIP = 'Skip';

var ActionListRow = function (_Component) {
  (0, _inherits3.default)(ActionListRow, _Component);

  function ActionListRow() {
    var _Object$getPrototypeO;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ActionListRow);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_Object$getPrototypeO = (0, _getPrototypeOf2.default)(ActionListRow)).call.apply(_Object$getPrototypeO, [this].concat(args))), _this), _this.state = { hover: false }, _this.shouldComponentUpdate = _function2.default, _this.handleMouseEnter = function () {
      _this.setState({ hover: true });
    }, _this.handleMouseLeave = function () {
      _this.setState({ hover: false });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ActionListRow, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      document.addEventListener('mouseleave', this.handleMouseLeave);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      document.removeEventListener('mouseleave', this.handleMouseLeave);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props;
      var styling = _props.styling;
      var isSelected = _props.isSelected;
      var action = _props.action;
      var isInitAction = _props.isInitAction;
      var onSelect = _props.onSelect;
      var timestamps = _props.timestamps;
      var isSkipped = _props.isSkipped;
      var hover = this.state.hover;

      var timeDelta = timestamps.current - timestamps.previous;
      var showButtons = hover && !isInitAction || isSkipped;

      var isButtonSelected = function isButtonSelected(btn) {
        return btn === BUTTON_SKIP && isSkipped;
      };

      return _react2.default.createElement(
        'div',
        (0, _extends3.default)({ onClick: onSelect,
          onMouseEnter: this.handleMouseEnter,
          onMouseLeave: this.handleMouseLeave
        }, styling(['actionListItem', isSelected && 'actionListItemSelected', isSkipped && 'actionListItemSkipped'], isSelected)),
        _react2.default.createElement(
          'div',
          styling(['actionListItemName', isSkipped && 'actionListItemNameSkipped']),
          action.type
        ),
        _react2.default.createElement(
          'div',
          styling('actionListItemButtons'),
          _react2.default.createElement(
            _RightSlider2.default,
            { styling: styling, shown: !showButtons, rotate: true },
            _react2.default.createElement(
              'div',
              styling('actionListItemTime'),
              timeDelta === 0 ? '+00:00:00' : (0, _dateformat2.default)(timeDelta, timestamps.previous ? '+MM:ss.L' : 'h:MM:ss.L')
            )
          ),
          _react2.default.createElement(
            _RightSlider2.default,
            { styling: styling, shown: showButtons, rotate: true },
            _react2.default.createElement(
              'div',
              styling('actionListItemSelector'),
              [BUTTON_SKIP].map(function (btn) {
                return (!isInitAction || btn !== BUTTON_SKIP) && _react2.default.createElement(
                  'div',
                  (0, _extends3.default)({ key: btn,
                    onClick: _this2.handleButtonClick.bind(_this2, btn)
                  }, styling(['selectorButton', isButtonSelected(btn) && 'selectorButtonSelected', 'selectorButtonSmall'], isButtonSelected(btn), true)),
                  btn
                );
              })
            )
          )
        )
      );
    }
  }, {
    key: 'handleButtonClick',
    value: function handleButtonClick(btn, e) {
      e.stopPropagation();

      switch (btn) {
        case BUTTON_SKIP:
          this.props.onViewClick();
          break;
      }
    }
  }]);
  return ActionListRow;
}(_react.Component);

ActionListRow.propTypes = {
  styling: _react.PropTypes.func.isRequired,
  isSelected: _react.PropTypes.bool.isRequired,
  action: _react.PropTypes.shape({ type: _react.PropTypes.string.isRequired }).isRequired,
  isInitAction: _react.PropTypes.bool.isRequired,
  onSelect: _react.PropTypes.func.isRequired,
  timestamps: _react.PropTypes.shape({
    current: _react.PropTypes.number.isRequired,
    previous: _react.PropTypes.number.isRequired
  }).isRequired,
  isSkipped: _react.PropTypes.bool.isRequired
};
exports.default = ActionListRow;
