// redux-devtools-inspector/lib/redux.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.updateMonitorState = updateMonitorState;
exports.reducer = reducer;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var UPDATE_MONITOR_STATE = '@@redux-devtools-inspector/UPDATE_MONITOR_STATE';

var DEFAULT_STATE = {
  isWideLayout: false,
  selectedActionId: null,
  startActionId: null,
  inspectedActionPath: [],
  inspectedStatePath: [],
  tabName: 'Diff',
  invertTheme: true,
  styling: function styling() {
    return {};
  }
};

function updateMonitorState(monitorState) {
  return { type: UPDATE_MONITOR_STATE, monitorState: monitorState };
}

function reduceUpdateState(state, action) {
  return action.type === UPDATE_MONITOR_STATE ? (0, _extends3.default)({}, state, action.monitorState) : state;
}

function reducer(props) {
  var state = arguments.length <= 1 || arguments[1] === undefined ? DEFAULT_STATE : arguments[1];
  var action = arguments[2];

  return (0, _extends3.default)({}, reduceUpdateState(state, action));
}
