// atom-languageclient/build/test/adapters/custom-linter-push-v2-adapter.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const linter_push_v2_adapter_1 = require("../../lib/adapters/linter-push-v2-adapter");
const ls = require("../../lib/languageclient");
const chai_1 = require("chai");
const atom_1 = require("atom");
const messageUrl = 'dummy';
const messageSolutions = ['dummy'];
class CustomLinterPushV2Adapter extends linter_push_v2_adapter_1.default {
    diagnosticToV2Message(path, diagnostic) {
        const message = super.diagnosticToV2Message(path, diagnostic);
        message.url = messageUrl;
        message.solutions = messageSolutions;
        return message;
    }
}
describe('CustomLinterPushV2Adapter', () => {
    describe('diagnosticToMessage', () => {
        it('converts Diagnostic and path to a linter$Message', () => {
            const filePath = '/a/b/c/d';
            const diagnostic = {
                message: 'This is a message',
                range: {
                    start: { line: 1, character: 2 },
                    end: { line: 3, character: 4 },
                },
                source: 'source',
                code: 'code',
                severity: ls.DiagnosticSeverity.Information,
            };
            const connection = { onPublishDiagnostics() { } };
            const adapter = new CustomLinterPushV2Adapter(connection);
            const result = adapter.diagnosticToV2Message(filePath, diagnostic);
            chai_1.expect(result.excerpt).equals(diagnostic.message);
            chai_1.expect(result.linterName).equals(diagnostic.source);
            chai_1.expect(result.location.file).equals(filePath);
            chai_1.expect(result.location.position).deep.equals(new atom_1.Range(new atom_1.Point(1, 2), new atom_1.Point(3, 4)));
            chai_1.expect(result.severity).equals('info');
            chai_1.expect(result.url).equals(messageUrl);
            chai_1.expect(result.solutions).deep.equals(messageSolutions);
        });
    });
});
