// redux-devtools-inspector/lib/tabs/StateTab.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactJsonTree = require('react-json-tree');

var _reactJsonTree2 = _interopRequireDefault(_reactJsonTree);

var _getItemString2 = require('./getItemString');

var _getItemString3 = _interopRequireDefault(_getItemString2);

var _getJsonTreeTheme = require('./getJsonTreeTheme');

var _getJsonTreeTheme2 = _interopRequireDefault(_getJsonTreeTheme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StateTab = function StateTab(_ref) {
  var nextState = _ref.nextState;
  var styling = _ref.styling;
  var base16Theme = _ref.base16Theme;
  var invertTheme = _ref.invertTheme;
  var labelRenderer = _ref.labelRenderer;
  return _react2.default.createElement(_reactJsonTree2.default, {
    labelRenderer: labelRenderer,
    theme: (0, _getJsonTreeTheme2.default)(base16Theme),
    data: nextState,
    getItemString: function getItemString(type, data) {
      return (0, _getItemString3.default)(styling, type, data);
    },
    invertTheme: invertTheme,
    hideRoot: true
  });
};

exports.default = StateTab;
