// remotedev-app/lib/constants/actionTypes.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var UPDATE_STATE = exports.UPDATE_STATE = 'devTools/UPDATE_STATE';
var SET_STATE = exports.SET_STATE = 'devTools/SET_STATE';
var SELECT_INSTANCE = exports.SELECT_INSTANCE = 'devTools/SELECT_INSTANCE';
var REMOVE_INSTANCE = exports.REMOVE_INSTANCE = 'devTools/REMOVE_INSTANCE';
var LIFTED_ACTION = exports.LIFTED_ACTION = 'devTools/LIFTED_ACTION';
var MONITOR_ACTION = exports.MONITOR_ACTION = 'devTools/MONITOR_ACTION';
var TOGGLE_SYNC = exports.TOGGLE_SYNC = 'devTools/TOGGLE_SYNC';
var SELECT_MONITOR = exports.SELECT_MONITOR = 'devTools/SELECT_MONITOR';
var TOGGLE_SLIDER = exports.TOGGLE_SLIDER = 'devTools/TOGGLE_SLIDER';
var TOGGLE_DISPATCHER = exports.TOGGLE_DISPATCHER = 'devTools/TOGGLE_DISPATCHER';
var SHOW_NOTIFICATION = exports.SHOW_NOTIFICATION = 'devTools/SHOW_NOTIFICATION';
var CLEAR_NOTIFICATION = exports.CLEAR_NOTIFICATION = 'devTools/CLEAR_NOTIFICATION';

var UPDATE_REPORTS = exports.UPDATE_REPORTS = 'reports/UPDATE';
var GET_REPORT_REQUEST = exports.GET_REPORT_REQUEST = 'reports/GET_REPORT_REQUEST';
var GET_REPORT_ERROR = exports.GET_REPORT_ERROR = 'reports/GET_REPORT_ERROR';
var GET_REPORT_SUCCESS = exports.GET_REPORT_SUCCESS = 'reports/GET_REPORT_SUCCESS';
var ERROR = exports.ERROR = 'ERROR';

var TEST_ADD = exports.TEST_ADD = 'test/ADD';
var TEST_EDIT = exports.TEST_EDIT = 'test/EDIT';
var TEST_REMOVE = exports.TEST_REMOVE = 'test/REMOVE';
var TEST_SELECT = exports.TEST_SELECT = 'test/SELECT';
