// javascript-typescript-langserver/lib/test/util-test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assert = require("assert");
const util_1 = require("../util");
describe('util', () => {
    describe('JSONPTR', () => {
        it('should escape JSON Pointer components', () => {
            const uri = 'file:///foo/~bar';
            const pointer = util_1.JSONPTR `/changes/${uri}/-`;
            assert.equal(pointer, '/changes/file:~1~1~1foo~1~0bar/-');
        });
    });
    describe('getMatchingPropertyCount()', () => {
        it('should return a score of 4 if 4 properties match', () => {
            const score = util_1.getMatchingPropertyCount({
                containerName: 'ts',
                kind: 'interface',
                name: 'Program',
                package: undefined,
            }, {
                containerKind: 'module',
                containerName: 'ts',
                kind: 'interface',
                name: 'Program',
                package: undefined,
            });
            assert.equal(score, 4);
        });
        it('should return a score of 0.6 if a string property is 60% similar', () => {
            const score = util_1.getMatchingPropertyCount({
                filePath: 'lib/foo.d.ts',
            }, {
                filePath: 'src/foo.ts',
            });
            assert.equal(score, 0.6);
        });
        it('should return a score of 4 if 4 properties match and 1 does not', () => {
            const score = util_1.getMatchingPropertyCount({
                containerKind: '',
                containerName: 'util',
                kind: 'var',
                name: 'colors',
                package: undefined,
            }, {
                containerKind: '',
                containerName: '',
                kind: 'var',
                name: 'colors',
                package: undefined,
            });
            assert.equal(score, 4);
        });
        it('should return a score of 3 if 3 properties match deeply', () => {
            const score = util_1.getMatchingPropertyCount({
                name: 'a',
                kind: 'class',
                package: { name: 'mypkg' },
                containerKind: undefined,
            }, {
                kind: 'class',
                name: 'a',
                containerKind: '',
                containerName: '',
                package: { name: 'mypkg' },
            });
            assert.equal(score, 3);
        });
    });
    describe('getPropertyCount()', () => {
        it('should return the amount of leaf properties', () => {
            const count = util_1.getPropertyCount({
                name: 'a',
                kind: 'class',
                package: {
                    name: 'mypkg',
                },
                containerKind: '',
            });
            assert.equal(count, 4);
        });
    });
    describe('isSymbolDescriptorMatch()', () => {
        it('should return true for a matching query', () => {
            const matches = util_1.isSymbolDescriptorMatch({
                containerKind: undefined,
                containerName: 'ts',
                kind: 'interface',
                name: 'Program',
                filePath: 'foo/bar.ts',
                package: undefined,
            }, {
                containerKind: 'module',
                containerName: 'ts',
                kind: 'interface',
                name: 'Program',
                filePath: 'foo/bar.ts',
                package: undefined,
            });
            assert.equal(matches, true);
        });
        it('should return true for a matching query with PackageDescriptor', () => {
            const matches = util_1.isSymbolDescriptorMatch({
                name: 'a',
                kind: 'class',
                package: { name: 'mypkg' },
                filePath: 'foo/bar.ts',
                containerKind: undefined,
            }, {
                kind: 'class',
                name: 'a',
                containerKind: '',
                containerName: '',
                filePath: 'foo/bar.ts',
                package: { name: 'mypkg' },
            });
            assert.equal(matches, true);
        });
    });
    describe('isGlobalTSFile()', () => {
        it('should match the synthetic reference to tsdlib when using importHelpers', () => {
            assert.equal(util_1.isGlobalTSFile('/node_modules/tslib/tslib.d.ts'), true);
        });
        it('should not include non-declaration files', () => {
            assert.equal(util_1.isGlobalTSFile('/node_modules/@types/node/Readme.MD'), false);
        });
        it('should include some libraries from @types with global declarations', () => {
            assert.equal(util_1.isGlobalTSFile('/node_modules/@types/node/index.d.ts'), true);
            assert.equal(util_1.isGlobalTSFile('/node_modules/@types/jest/index.d.ts'), true);
            assert.equal(util_1.isGlobalTSFile('/node_modules/@types/jasmine/index.d.ts'), true);
            assert.equal(util_1.isGlobalTSFile('/node_modules/@types/mocha/index.d.ts'), true);
        });
    });
    describe('path2uri()', () => {
        it('should throw an error if a non-absolute uri is passed in', () => {
            assert.throws(() => util_1.path2uri('baz/qux'));
        });
        it('should convert a Unix file path to a URI', () => {
            const uri = util_1.path2uri('/baz/qux');
            assert.equal(uri, 'file:///baz/qux');
        });
        it('should convert a Windows file path to a URI', () => {
            const uri = util_1.path2uri('C:\\baz\\qux');
            assert.equal(uri, 'file:///C:/baz/qux');
        });
        it('should encode special characters', () => {
            const uri = util_1.path2uri('/💩');
            assert.equal(uri, 'file:///%F0%9F%92%A9');
        });
        it('should encode unreserved special characters', () => {
            const uri = util_1.path2uri('/@baz');
            assert.equal(uri, 'file:///%40baz');
        });
    });
    describe('uri2path()', () => {
        it('should convert a Unix file URI to a file path', () => {
            const filePath = util_1.uri2path('file:///baz/qux');
            assert.equal(filePath, '/baz/qux');
        });
        it('should convert a Windows file URI to a file path', () => {
            const filePath = util_1.uri2path('file:///c:/baz/qux');
            assert.equal(filePath, 'c:\\baz\\qux');
        });
        it('should convert a Windows file URI with uppercase drive letter to a file path', () => {
            const filePath = util_1.uri2path('file:///C:/baz/qux');
            assert.equal(filePath, 'C:\\baz\\qux');
        });
        it('should decode special characters', () => {
            const filePath = util_1.uri2path('file:///%F0%9F%92%A9');
            assert.equal(filePath, '/💩');
        });
        it('should decode unreserved special characters', () => {
            const filePath = util_1.uri2path('file:///%40foo');
            assert.equal(filePath, '/@foo');
        });
    });
});
