// vscode-chrome-debug-core/out/src/transformers/basePathTransformer.js
"use strict";
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Converts a local path from Code to a path on the target.
 */
class BasePathTransformer {
    launch(args) {
        return Promise.resolve();
    }
    attach(args) {
        return Promise.resolve();
    }
    setBreakpoints(args) {
    }
    clearTargetContext() {
    }
    scriptParsed(scriptPath) {
        return scriptPath;
    }
    breakpointResolved(bp, targetPath) {
        return this.getClientPathFromTargetPath(targetPath) || targetPath;
    }
    stackTraceResponse(response) {
    }
    fixSource(source) {
    }
    getTargetPathFromClientPath(clientPath) {
        return clientPath;
    }
    getClientPathFromTargetPath(targetPath) {
        return targetPath;
    }
}
exports.BasePathTransformer = BasePathTransformer;
