// rebass/dist/themes/biblio.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _basic = require('./basic');

var _basic2 = _interopRequireDefault(_basic);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var sans = {
  fontFamily: 'Roboto, sans-serif'
};
var caps = {
  textTransform: 'uppercase',
  letterSpacing: '.1em'
};

var colors = {
  red: '#e54',
  blue: '#059',
  green: '#0b7',
  midgray: '#444',
  gray: '#eee'
};

var biblio = {
  name: 'Biblio',
  fontFamily: 'Palatino, Georgia, serif',
  color: '#111',
  backgroundColor: '#fff',

  colors: _extends({}, _basic2.default.colors, colors, {
    primary: colors.red,
    error: colors.red,
    warning: _basic2.default.colors.orange,
    info: colors.blue,
    success: colors.green,
    secondary: colors.midgray,
    default: colors.midgray,
    muted: colors.gray
  }),

  borderColor: 'rgba(0, 0, 0, ' + 1 / 8 + ')',

  scale: [0, 12, 24, 48, 96],

  fontSizes: [72, 64, 48, 32, 18, 16, 14],

  bold: 500,

  Heading: {
    alt: _extends({
      opacity: 1
    }, sans, caps, {
      fontSize: 14,
      color: colors.red
    })
  },

  Banner: {
    backgroundColor: '#f6fee6',
    // backgroundImage: 'none',
    boxShadow: 'inset 0 0 320px 0 rgba(128, 64, 0, .5), inset 0 0 0 99999px rgba(128, 128, 96, .25)'
  },

  Toolbar: {
    color: 'inherit',
    backgroundColor: '#fff',
    borderBottom: '1px solid rgba(0, 0, 0, ' + 1 / 8 + ')'
  },

  Button: _extends({}, sans, caps, {
    fontSize: 12
  }),
  ButtonOutline: _extends({}, sans, caps, {
    fontSize: 12
  }),
  NavItem: _extends({}, sans, caps, {
    fontSize: 12
  }),
  PanelHeader: _extends({}, sans),
  Label: _extends({}, sans, caps, {
    fontSize: 12,
    opacity: 5 / 8
  }),
  SequenceMap: _extends({}, sans, caps, {
    fontSize: 12
  }),
  Donut: _extends({}, sans),
  Stat: _extends({}, sans),
  Breadcrumbs: _extends({}, sans, {
    color: '#e54'
  }),

  PageHeader: {
    borderColor: '#e54'
  },
  SectionHeader: {
    borderColor: '#e54'
  },
  Rating: {
    fontSize: 24
  }
};

exports.default = biblio;
