// rebass/dist/themes/dark.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _basic = require('./basic');

var _basic2 = _interopRequireDefault(_basic);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var caps = {
  textTransform: 'uppercase',
  letterSpacing: '.2em'
};

var dark = {
  name: 'Dark',
  fontFamily: '"Avenir Next", Helvetica, sans-serif',
  color: '#eee',
  backgroundColor: '#111',
  borderRadius: 5,
  borderColor: _basic2.default.colors.lighten,

  colors: _extends({}, _basic2.default.colors, {
    blue: '#00d9f7',
    info: '#00d9f7',
    green: '#0f8',
    success: '#0f8',
    orange: '#fb3',
    warning: '#fb3',
    primary: '#778',
    midgray: '#778',
    gray: '#333339',
    secondary: '#333339',
    red: '#f04',
    error: '#f04'
  }),
  inverted: '#222',

  scale: [0, 10, 20, 40, 80],

  Card: {
    backgroundColor: '#222',
    border: 0
  },
  Panel: {
    borderWidth: 2,
    backgroundColor: '#000'
  },
  NavItem: _extends({
    fontSize: 12
  }, caps),
  Heading: _extends({}, caps),
  Button: _extends({}, caps, {
    fontSize: 12,
    color: '#00d9f7',
    opacity: 7 / 8
  }),
  ButtonOutline: _extends({}, caps, {
    fontSize: 12,
    color: '#00d9f7',
    opacity: 7 / 8
  }),
  Toolbar: {
    minHeight: 64,
    color: '#00d9f7',
    backgroundColor: 'rgba(0, 0, 0, ' + 7 / 8 + ')'
  },
  Label: { opacity: 5 / 8 },
  Menu: {
    color: '#00d9f7',
    backgroundColor: '#000'
  },
  Message: {
    color: '#111',
    opacity: 15 / 16
  },
  Text: {
    opacity: 7 / 8
  },
  Footer: {
    opacity: 1 / 2
  }
};

exports.default = dark;
