// rebass/dist/themes/hello.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _basic = require('./basic');

var _basic2 = _interopRequireDefault(_basic);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var borderWidth = 4;
var borderColor = '#346';

var colors = {
  black: '#346',
  blue: '#0df',
  red: '#f7b',
  green: '#3ea',
  orange: '#fe1',
  gray: '#eef',
  midgray: '#ccd'
};

var hello = {
  name: 'Hello',
  fontFamily: 'Karla, sans-serif',
  letterSpacing: '.1em',
  color: '#346',
  backgroundColor: '#fff',

  borderRadius: 6,

  colors: _extends({}, _basic2.default.colors, colors, {
    default: colors.midgray,
    primary: colors.blue,
    secondary: colors.midgray,
    info: colors.blue,
    error: colors.blue,
    success: colors.green,
    warning: colors.orange
  }),
  inverted: colors.black,

  scale: [0, 8, 16, 32, 64],

  bold: 700,

  Toolbar: {
    minHeight: 64,
    color: colors.black,
    backgroundColor: '#fff'
  },
  Banner: {
    backgroundImage: 'none',
    color: colors.black,
    backgroundColor: colors.blue
  },
  Button: {
    fontSize: 20,
    paddingTop: 16,
    paddingBottom: 16
  },
  ButtonOutline: {
    backgroundColor: '#346'
  },
  ButtonCircle: {
    fontSize: 16,
    lineHeight: '16px',
    width: 48,
    height: 48
  },
  Panel: {
    borderWidth: borderWidth
  },
  PanelFooter: {
    borderWidth: borderWidth
  },
  PageHeader: {
    borderWidth: borderWidth,
    borderColor: borderColor
  },
  SectionHeader: {
    borderWidth: borderWidth,
    borderColor: borderColor
  },
  Footer: {
    borderWidth: borderWidth,
    borderColor: borderColor
  },
  Input: {
    input: {
      fontSize: 20,
      height: 48,
      paddingTop: 16,
      paddingBottom: 16
    }
  },
  Select: {
    select: {
      fontSize: 20,
      height: 48
    },
    arrow: {
      marginTop: 24,
      marginBottom: 24
    }
  },
  Checkbox: {
    box: {
      width: 24,
      height: 24
    }
  },
  Radio: {
    dot: {
      width: 24,
      height: 24
    }
  },
  Rating: {
    fontSize: 24
  }
};

exports.default = hello;
