// rebass/dist/themes/mono.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _basic = require('./basic');

var _basic2 = _interopRequireDefault(_basic);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var caps = {
  textTransform: 'uppercase',
  letterSpacing: '.1em'
};

var mono = {
  name: 'Mono',
  fontFamily: 'Roboto, sans-serif',
  color: '#111',
  backgroundColor: '#fff',

  scale: [0, 8, 16, 32, 64],

  fontSizes: [64, 24, 20, 18, 14, 12, 12],

  borderColor: _basic2.default.colors.darken,

  colors: _extends({}, _basic2.default.colors, {
    primary: '#111'
  }),

  Heading: _extends({
    fontFamily: 'inherit'
  }, caps),

  Button: _extends({}, caps),
  NavItem: _extends({}, caps),
  PageHeader: { borderColor: '#111' },
  Footer: { borderColor: '#111' },
  Checkbox: {
    box: {
      width: 24,
      height: 24
    }
  },
  Radio: {
    dot: {
      width: 24,
      height: 24
    }
  },
  Pre: {
    fontSize: 14,
    borderLeft: 0,
    padding: 8,
    borderRadius: 2,
    backgroundColor: _basic2.default.colors.gray
  }
};

exports.default = mono;
