// rebass/dist/Avatar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * A circular image for displaying user avatars
 */

var Avatar = function Avatar(_ref, _ref2) {
  var size = _ref.size;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['size', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var rebass = _ref2.rebass;
  var colors = theme.colors;


  var cx = (0, _classnames2.default)('Avatar', className);

  var sx = _extends({
    display: 'inline-block',
    maxWidth: 'none',
    width: size,
    height: size,
    backgroundColor: colors.gray,
    borderRadius: 99999
  }, style);

  return _react2.default.createElement('img', _extends({}, props, {
    className: cx,
    width: size,
    height: size,
    style: sx
  }));
};

Avatar.propTypes = {
  /** Width and height of image in pixels */
  size: _react2.default.PropTypes.number
};

Avatar.defaultProps = {
  size: 48
};

Avatar._name = 'Avatar';

exports.default = (0, _withRebass2.default)(Avatar);
