// rebass/dist/Badge.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _baseline = require('./util/baseline');

var _baseline2 = _interopRequireDefault(_baseline);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/** Component for displaying small status indicators */

var Badge = function Badge(_ref) {
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['className', 'style', 'theme', 'subComponentStyles']);

  var fontSizes = theme.fontSizes;
  var bold = theme.bold;
  var baselineShift = theme.baselineShift;
  var scale = theme.scale;
  var colors = theme.colors;
  var borderRadius = theme.borderRadius;


  var cx = (0, _classnames2.default)('Badge', className);

  var pad = (0, _baseline2.default)(baselineShift)(scale[1]);
  var xpad = props.circle ? 0 : scale[1];

  var sx = _extends({
    fontSize: fontSizes[6],
    fontWeight: bold,
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    verticalAlignment: 'baseline',
    width: props.circle ? scale[2] : null,
    height: scale[2]
  }, pad, {
    paddingLeft: xpad,
    paddingRight: xpad,
    overflow: 'hidden',
    borderRadius: borderRadius,
    color: colors.white,
    backgroundColor: colors.default
  }, style.fill, style);

  return _react2.default.createElement('div', _extends({}, props, {
    className: cx,
    style: sx }));
};

Badge._name = 'Badge';

exports.default = (0, _withRebass2.default)(Badge);
