// rebass/dist/CardImage.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Image for use within the Card component
 */

var CardImage = function CardImage(_ref) {
  var src = _ref.src;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['src', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;


  var cx = (0, _classnames2.default)('CardImage', className);

  var sx = _extends({
    display: 'block',
    width: 'calc(100% + ' + 2 * scale[1] + 'px)',
    maxWidth: 'none',
    height: 'auto',
    margin: -scale[1],
    marginBottom: scale[1]
  }, style);

  return _react2.default.createElement('img', _extends({}, props, {
    src: src,
    className: cx,
    style: sx }));
};

CardImage.propTypes = {
  /** Image source */
  src: _react2.default.PropTypes.string.isRequired
};

CardImage._name = 'CardImage';

exports.default = (0, _withRebass2.default)(CardImage);
