// rebass/dist/Carousel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Carousel that wraps children and displays one at a time
 */

var Carousel = function Carousel(_ref) {
  var index = _ref.index;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['index', 'className', 'style', 'theme', 'subComponentStyles']);

  var length = props.children ? props.children.length || 0 : 0;

  var cx = (0, _classnames2.default)('Carousel', className);

  var sx = {
    root: _extends({
      overflow: 'hidden'
    }, style),
    inner: _extends({
      whiteSpace: 'nowrap',
      height: '100%',
      transition: 'transform .2s ease-out',
      transform: 'translateX(' + index % length * -100 + '%)'
    }, subComponentStyles.inner),
    child: _extends({
      display: 'inline-block',
      verticalAlign: 'middle',
      whiteSpace: 'normal',
      width: '100%',
      height: '100%'
    }, subComponentStyles.child)
  };

  var children = _react2.default.Children.map(props.children, function (child, i) {
    return _react2.default.createElement(
      'div',
      { style: sx.child },
      child
    );
  });

  return _react2.default.createElement(
    'div',
    _extends({}, props, {
      className: cx,
      style: sx.root }),
    _react2.default.createElement(
      'div',
      { style: sx.inner },
      children
    )
  );
};

Carousel.propTypes = {
  /** Index for the child to show */
  index: _react2.default.PropTypes.number
};

Carousel._name = 'Carousel';

exports.default = (0, _withRebass2.default)(Carousel);
