// rebass/dist/Close.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * A button with an × for close and dismiss actions
 */

var Close = function Close(_ref) {
  var baseRef = _ref.baseRef;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['baseRef', 'className', 'style', 'theme', 'subComponentStyles']);

  var cx = (0, _classnames2.default)('Close', className);

  var sx = _extends({
    fontSize: '1.5em',
    lineHeight: 1,
    fontWeight: 'bold',
    margin: 0,
    padding: 0,
    cursor: 'pointer',
    color: 'inherit',
    backgroundColor: 'transparent',
    border: 0,
    WebkitAppearance: 'none'
  }, style);

  return _react2.default.createElement('button', _extends({}, props, {
    ref: baseRef,
    title: 'Close',
    className: cx,
    style: sx,
    children: '\xD7' }));
};

Close.propTypes = {
  /** Adds a ref to the button element */
  baseRef: _react2.default.PropTypes.func
};

Close._name = 'Close';

exports.default = (0, _withRebass2.default)(Close);
