// rebass/dist/Heading.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Heading element with no margin and size based on fontSizes scale
 */

var Heading = function Heading(_ref) {
  var level = _ref.level;
  var size = _ref.size;
  var big = _ref.big;
  var alt = _ref.alt;
  var center = _ref.center;
  var className = _ref.className;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;
  var style = _ref.style;

  var props = _objectWithoutProperties(_ref, ['level', 'size', 'big', 'alt', 'center', 'className', 'theme', 'subComponentStyles', 'style']);

  var fontSizes = theme.fontSizes;
  var bold = theme.bold;

  var Comp = 'h' + level;

  var h = function h(n) {
    return fontSizes[n];
  };

  var fontSize = typeof size === 'number' ? h(size) : h(level);

  if (big) {
    fontSize *= 2;
  }

  var cx = (0, _classnames2.default)('Heading', className, {
    'Heading_alt': alt
  });

  var sx = _extends({
    fontSize: fontSize,
    fontWeight: bold,
    lineHeight: 1.25,
    textAlign: center ? 'center' : null,
    margin: 0
  }, alt ? subComponentStyles.alt : {}, style);

  return _react2.default.createElement(Comp, _extends({}, props, {
    className: cx,
    style: sx }));
};

Heading.propTypes = {
  /** Doubles the visual size - useful for marketing pages */
  big: _react2.default.PropTypes.bool,
  /** Heading level, e.g. level={1} for <h1> */
  level: _react2.default.PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /** Visual size of heading */
  size: _react2.default.PropTypes.oneOf([0, 1, 2, 3, 4, 5, 6]),
  /** Applies alternate styling - useful for slugs and subheadings */
  alt: _react2.default.PropTypes.bool,
  /** Centers text alignment */
  center: _react2.default.PropTypes.bool
};

Heading.defaultProps = {
  level: 2
};

Heading._name = 'Heading';

exports.default = (0, _withRebass2.default)(Heading);
