// rebass/dist/HeadingLink.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _Heading = require('./Heading');

var _Heading2 = _interopRequireDefault(_Heading);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Heading element with unstyled link. Useful for in-page navigation
 */

var HeadingLink = function HeadingLink(_ref) {
  var level = _ref.level;
  var size = _ref.size;
  var href = _ref.href;
  var alt = _ref.alt;
  var style = _ref.style;
  var className = _ref.className;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['level', 'size', 'href', 'alt', 'style', 'className', 'theme', 'subComponentStyles']);

  var cx = (0, _classnames2.default)('HeadingLink', className);

  var sx = {
    root: _extends({}, style),
    link: _extends({
      color: 'inherit',
      textDecoration: 'none'
    }, subComponentStyles.link)
  };

  return _react2.default.createElement(
    _Heading2.default,
    {
      alt: alt,
      level: level,
      size: size,
      className: cx,
      style: sx.root },
    _react2.default.createElement('a', _extends({}, props, {
      href: href,
      style: sx.link }))
  );
};

HeadingLink.propTypes = {
  /** Heading level, e.g. level={1} for <h1> */
  level: _react2.default.PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /** Visual size of heading */
  size: _react2.default.PropTypes.oneOf([1, 2, 3, 4, 5, 6]),
  /** href for link */
  href: _react2.default.PropTypes.string
};

HeadingLink.defaultProps = {
  level: 2,
  href: '#!'
};

HeadingLink._name = 'HeadingLink';

exports.default = (0, _withRebass2.default)(HeadingLink);
