// rebass/dist/Input.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _Label = require('./Label');

var _Label2 = _interopRequireDefault(_Label);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Input element with label with support for aria-invalid, disabled, and readOnly HTML attributes
 */

var Input = function Input(_ref) {
  var label = _ref.label;
  var name = _ref.name;
  var type = _ref.type;
  var message = _ref.message;
  var hideLabel = _ref.hideLabel;
  var horizontal = _ref.horizontal;
  var children = _ref.children;
  var autoOff = _ref.autoOff;
  var baseRef = _ref.baseRef;
  var ref = _ref.ref;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['label', 'name', 'type', 'message', 'hideLabel', 'horizontal', 'children', 'autoOff', 'baseRef', 'ref', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;
  var colors = theme.colors;
  var borderColor = theme.borderColor;


  var invalid = props.invalid || props['aria-invalid'];

  var cx = (0, _classnames2.default)('Input', className, {
    'isInvalid': invalid,
    'isDisabled': props.disabled,
    'isReadonly': props.readOnly
  });

  var _style$borderRadius = style.borderRadius;
  var borderRadius = _style$borderRadius === undefined ? theme.borderRadius : _style$borderRadius;
  var color = style.color;
  var backgroundColor = style.backgroundColor;

  var rootStyle = _objectWithoutProperties(style, ['borderRadius', 'color', 'backgroundColor']);

  var sx = {
    root: _extends({
      display: horizontal ? 'flex' : null,
      alignItems: horizontal ? 'baseline' : null,
      marginBottom: scale[2],
      color: invalid ? colors.error : null
    }, rootStyle),
    label: _extends({
      minWidth: horizontal ? 96 : null,
      paddingRight: horizontal ? scale[1] : null
    }, subComponentStyles.label),
    input: _extends({
      fontFamily: 'inherit',
      fontSize: 'inherit',
      boxSizing: 'border-box',
      display: 'block',
      flex: horizontal ? '1 1 auto' : null,
      width: '100%',
      height: scale[3],
      margin: 0,
      paddingLeft: scale[1],
      paddingRight: scale[1],
      color: color || 'inherit',
      backgroundColor: backgroundColor || colors.lighten,
      borderWidth: 1,
      borderStyle: 'solid',
      borderColor: invalid ? colors.error : borderColor,
      borderRadius: borderRadius,
      boxShadow: 'none',
      WebkitAppearance: 'none',
      appearance: 'none'
    }, style.fill, subComponentStyles.input),
    message: _extends({
      paddingLeft: horizontal ? scale[1] : null
    }, subComponentStyles.message)
  };

  var autoProps = autoOff ? {
    autoComplete: 'off',
    autoCorrect: 'off',
    autoCapitalize: 'off',
    spellCheck: 'off'
  } : {};

  return _react2.default.createElement(
    'div',
    {
      className: cx,
      style: sx.root },
    _react2.default.createElement(_Label2.default, {
      htmlFor: name,
      hide: hideLabel,
      style: sx.label,
      children: label }),
    _react2.default.createElement('input', _extends({}, props, autoProps, {
      ref: baseRef,
      type: type,
      name: name,
      style: sx.input })),
    message && _react2.default.createElement(_Text2.default, {
      small: true,
      style: sx.message,
      children: message })
  );
};

Input.propTypes = {
  /** Label for form element */
  label: _react2.default.PropTypes.string.isRequired,
  /** Name attribute for form element */
  name: _react2.default.PropTypes.string.isRequired,
  /** Form element type */
  type: _react2.default.PropTypes.string,
  /** Adds a helper or error message below the input */
  message: _react2.default.PropTypes.string,
  /** Hides the form element label */
  hideLabel: _react2.default.PropTypes.bool,
  /** Displays label to the left */
  horizontal: _react2.default.PropTypes.bool,
  /** Disables autocomplete, autocorrect, autocapitalize, and spellcheck props */
  autoOff: _react2.default.PropTypes.bool,
  /** Adds a ref to the input element */
  baseRef: _react2.default.PropTypes.func
};

Input.defaultProps = {
  type: 'text'
};

Input._name = 'Input';

exports.default = (0, _withRebass2.default)(Input);
