// rebass/dist/Media.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Media object with vertical alignment using flexbox
 */

var Media = function Media(_ref) {
  var img = _ref.img;
  var right = _ref.right;
  var align = _ref.align;
  var imgWidth = _ref.imgWidth;
  var imgHeight = _ref.imgHeight;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['img', 'right', 'align', 'imgWidth', 'imgHeight', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;


  var alignment = {
    top: 'flex-start',
    center: 'center',
    bottom: 'flex-end'
  };

  var alignItems = alignment[align];

  var cx = (0, _classnames2.default)('Media', className);

  var sx = {
    root: _extends({
      display: 'flex',
      marginBottom: scale[2],
      alignItems: alignItems
    }, style),
    image: _extends({
      flex: 'none',
      maxWidth: 'none',
      marginRight: right ? 0 : scale[2],
      marginLeft: right ? scale[2] : 0,
      order: right ? 9999 : null
    }, subComponentStyles.image),
    body: _extends({}, subComponentStyles.body)
  };

  return _react2.default.createElement(
    'div',
    _extends({}, props, {
      className: cx,
      style: sx.root }),
    _react2.default.createElement('img', {
      src: img,
      width: imgWidth,
      height: imgHeight,
      style: sx.image }),
    _react2.default.createElement('div', {
      style: sx.body,
      children: children })
  );
};

Media.propTypes = {
  /** Image source */
  img: _react2.default.PropTypes.string,
  /** Displays image to the right */
  right: _react2.default.PropTypes.bool,
  /** Vertical alignment */
  align: _react2.default.PropTypes.oneOf(['top', 'center', 'bottom'])
};

Media._name = 'Media';

exports.default = (0, _withRebass2.default)(Media);
