// rebass/dist/PageHeader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _Heading = require('./Heading');

var _Heading2 = _interopRequireDefault(_Heading);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Main page header with description
 */

var PageHeader = function PageHeader(_ref) {
  var heading = _ref.heading;
  var description = _ref.description;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['heading', 'description', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;
  var borderColor = theme.borderColor;


  var cx = (0, _classnames2.default)('PageHeader', className);

  var sx = {
    root: _extends({
      display: 'flex',
      flexWrap: 'wrap',
      alignItems: 'center',
      paddingTop: scale[3],
      paddingBottom: scale[2],
      marginTop: scale[4],
      marginBottom: scale[4],
      borderBottomWidth: 2,
      borderBottomStyle: 'solid',
      borderColor: borderColor
    }, style),
    inner: _extends({
      flex: '1 1 auto'
    }, subComponentStyles.inner),
    heading: _extends({}, subComponentStyles.Heading),
    text: _extends({}, subComponentStyles.Text)
  };

  return _react2.default.createElement(
    'header',
    _extends({}, props, {
      className: cx,
      style: sx.root }),
    _react2.default.createElement(
      'div',
      { style: sx.inner },
      _react2.default.createElement(_Heading2.default, {
        level: 1,
        style: sx.heading,
        children: heading }),
      description && _react2.default.createElement(_Text2.default, {
        style: sx.text,
        children: description })
    ),
    children
  );
};

PageHeader.propTypes = {
  /** Page heading */
  heading: _react2.default.PropTypes.string,
  /** Description of page */
  description: _react2.default.PropTypes.string
};

PageHeader._name = 'PageHeader';

exports.default = (0, _withRebass2.default)(PageHeader);
