// rebass/dist/PanelFooter.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Footer for Panel component with vertical centering using flexbox
 */

var PanelFooter = function PanelFooter(_ref) {
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;
  var borderRadius = theme.borderRadius;
  var fontSizes = theme.fontSizes;


  var cx = (0, _classnames2.default)('PanelFooter', className);

  var sx = _extends({
    fontSize: fontSizes[6],
    display: 'flex',
    alignItems: 'center',
    marginTop: scale[2],
    marginRight: -scale[2],
    marginBottom: -scale[2],
    marginLeft: -scale[2],
    padding: scale[2],
    borderTopWidth: 1,
    borderTopStyle: 'solid',
    borderRadius: '0 0 ' + borderRadius + 'px ' + borderRadius + 'px'
  }, style.fill, style);

  return _react2.default.createElement('div', _extends({}, props, {
    className: cx,
    style: sx }));
};

PanelFooter._name = 'PanelFooter';

exports.default = (0, _withRebass2.default)(PanelFooter);
