// rebass/dist/SectionHeader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _HeadingLink = require('./HeadingLink');

var _HeadingLink2 = _interopRequireDefault(_HeadingLink);

var _Text = require('./Text');

var _Text2 = _interopRequireDefault(_Text);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Header for section elements
 */

var SectionHeader = function SectionHeader(_ref) {
  var heading = _ref.heading;
  var href = _ref.href;
  var description = _ref.description;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['heading', 'href', 'description', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;
  var borderColor = theme.borderColor;


  var cx = (0, _classnames2.default)('SectionHeader', className);

  var sx = {
    root: _extends({
      display: 'flex',
      alignItems: 'center',
      paddingBottom: scale[1],
      marginTop: scale[3],
      marginBottom: scale[3],
      borderBottomWidth: 1,
      borderBottomStyle: 'solid',
      borderBottomColor: borderColor
    }, style),
    inner: _extends({
      flex: '1 1 auto'
    }, subComponentStyles.inner),
    Heading: _extends({}, subComponentStyles.Heading),
    Text: _extends({}, subComponentStyles.Text)
  };

  href = href || '#' + (heading || '');

  return _react2.default.createElement(
    'header',
    _extends({}, props, {
      className: cx,
      style: sx.root }),
    _react2.default.createElement(
      'div',
      { style: sx.inner },
      _react2.default.createElement(_HeadingLink2.default, {
        href: href,
        style: sx.Heading,
        children: heading }),
      description && _react2.default.createElement(_Text2.default, {
        style: sx.Text,
        children: description })
    ),
    children
  );
};

SectionHeader.propTypes = {
  /** Section heading */
  heading: _react2.default.PropTypes.string,
  /** Link to section, used in HeadingLink */
  href: _react2.default.PropTypes.string,
  /** Description of section */
  description: _react2.default.PropTypes.string
};

SectionHeader._name = 'SectionHeader';

exports.default = (0, _withRebass2.default)(SectionHeader);
