// rebass/dist/SequenceMap.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

var _SequenceMapStep = require('./SequenceMapStep');

var _SequenceMapStep2 = _interopRequireDefault(_SequenceMapStep);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Sequence map pattern for use in multi-step forms
 */

var SequenceMap = function SequenceMap(_ref) {
  var steps = _ref.steps;
  var active = _ref.active;
  var children = _ref.children;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['steps', 'active', 'children', 'className', 'style', 'theme', 'subComponentStyles']);

  var fontSizes = theme.fontSizes;
  var bold = theme.bold;
  var colors = theme.colors;


  var cx = (0, _classnames2.default)('SequenceMap', className);

  var sx = {
    root: _extends({
      display: 'flex',
      alignItems: 'flex-start',
      justifyContent: 'space-between',
      fontSize: fontSizes[5],
      fontWeight: bold,
      color: colors.gray
    }, style),
    step: _extends({}, subComponentStyles.step)
  };

  var chx = _react2.default.Children.map(children, function (child, i) {
    return _react2.default.cloneElement(child, {
      width: 1 / children.length * 100 + '%',
      first: i === 0,
      style: sx.step
    });
  });

  var schx = steps.map(function (step, i) {
    return _react2.default.createElement(_SequenceMapStep2.default, _extends({
      key: i,
      first: i === 0,
      width: 100 / steps.length + '%',
      active: i <= active,
      style: sx.step
    }, step));
  });

  return _react2.default.createElement('div', _extends({}, props, {
    children: chx || schx,
    className: cx,
    style: sx.root }));
};

SequenceMap.propTypes = {
  /** Array of links for each step in the sequence */
  steps: _react2.default.PropTypes.array,
  /** Index of current step */
  active: _react2.default.PropTypes.number
};

SequenceMap.defaultProps = {
  steps: []
};

SequenceMap.Step = _SequenceMapStep2.default;

SequenceMap._name = 'SequenceMap';

exports.default = (0, _withRebass2.default)(SequenceMap);
