// rebass/dist/TabItem.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var TabItem = function TabItem(_ref) {
  var _ref$is = _ref.is;
  var is = _ref$is === undefined ? 'a' : _ref$is;
  var active = _ref.active;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['is', 'active', 'className', 'style', 'theme', 'subComponentStyles']);

  var scale = theme.scale;
  var fontSizes = theme.fontSizes;
  var bold = theme.bold;
  var colors = theme.colors;

  var Comp = is;

  var cx = (0, _classnames2.default)('TabItem', className);

  var sx = _extends({
    fontWeight: bold,
    fontSize: fontSizes[5],
    paddingTop: scale[1],
    paddingBottom: scale[1],
    marginRight: scale[2],
    textDecoration: 'none',
    borderBottomWidth: 2,
    borderBottomStyle: 'solid',
    borderBottomColor: 'transparent',
    cursor: 'pointer'
  }, active ? {
    color: colors.primary,
    borderBottomColor: colors.primary
  } : {}, style);

  return _react2.default.createElement(Comp, _extends({}, props, {
    className: cx,
    style: sx }));
};

TabItem._name = 'TabItem';

exports.default = (0, _withRebass2.default)(TabItem);
