// rebass/dist/Text.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _withRebass = require('./withRebass');

var _withRebass2 = _interopRequireDefault(_withRebass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

/**
 * Component for displaying text in UI
 */

var Text = function Text(_ref) {
  var size = _ref.size;
  var small = _ref.small;
  var className = _ref.className;
  var style = _ref.style;
  var theme = _ref.theme;
  var subComponentStyles = _ref.subComponentStyles;

  var props = _objectWithoutProperties(_ref, ['size', 'small', 'className', 'style', 'theme', 'subComponentStyles']);

  var fontSizes = theme.fontSizes;


  var cx = (0, _classnames2.default)('Text', className);

  size = size || (small ? 6 : 4);
  var sx = _extends({
    fontSize: fontSizes[size],
    margin: 0
  }, style);

  return _react2.default.createElement('p', _extends({}, props, {
    className: cx,
    style: sx }));
};

Text.propTypes = {
  /** Sets a smaller font size */
  small: _react2.default.PropTypes.bool
};

Text._name = 'Text';

exports.default = (0, _withRebass2.default)(Text);
