// flow-language-server/lib/pkg/nuclide-language-service-rpc/lib/NullLanguageService.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NullLanguageService = undefined;

var _rxjs = require('rxjs');

// An implementation of LanguageService which always returns no results.
// Useful for implementing aggregate language services.
/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 * @format
 */

class NullLanguageService {
  getDiagnostics(fileVersion) {
    return Promise.resolve(null);
  }

  observeDiagnostics() {
    return _rxjs.Observable.empty().publish();
  }

  getAutocompleteSuggestions(fileVersion, position, activatedManually, prefix) {
    return Promise.resolve(null);
  }

  getDefinition(fileVersion, position) {
    return Promise.resolve(null);
  }

  findReferences(fileVersion, position) {
    return Promise.resolve(null);
  }

  getCoverage(filePath) {
    return Promise.resolve(null);
  }

  getOutline(fileVersion) {
    return Promise.resolve(null);
  }

  typeHint(fileVersion, position) {
    return Promise.resolve(null);
  }

  highlight(fileVersion, position) {
    return Promise.resolve(null);
  }

  formatSource(fileVersion, range) {
    return Promise.resolve(null);
  }

  formatEntireFile(fileVersion, range) {
    return Promise.resolve(null);
  }

  formatAtPosition(fileVersion, position, triggerCharacter) {
    return Promise.resolve(null);
  }

  getEvaluationExpression(fileVersion, position) {
    return Promise.resolve(null);
  }

  supportsSymbolSearch(directories) {
    return Promise.resolve(false);
  }

  symbolSearch(query, directories) {
    return Promise.resolve(null);
  }

  getProjectRoot(fileUri) {
    return Promise.resolve(null);
  }

  isFileInProject(fileUri) {
    return Promise.resolve(false);
  }

  dispose() {}
}
exports.NullLanguageService = NullLanguageService;
