// flow-language-server/lib/pkg/nuclide-open-files-rpc/lib/main.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OPEN_FILES_SERVICE = exports.ConfigObserver = exports.FileEventKind = exports.FileVersionNotifier = exports.FileCache = undefined;

var _constants = require('./constants');

Object.defineProperty(exports, 'FileEventKind', {
  enumerable: true,
  get: function get() {
    return _constants.FileEventKind;
  }
});

var _ConfigObserver = require('./ConfigObserver');

Object.defineProperty(exports, 'ConfigObserver', {
  enumerable: true,
  get: function get() {
    return _ConfigObserver.ConfigObserver;
  }
});
exports.getBufferAtVersion = getBufferAtVersion;

var _FileCache = require('./FileCache');

var _FileVersionNotifier = require('./FileVersionNotifier');

var _assert = require('assert');

var _assert2 = _interopRequireDefault(_assert);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.FileCache = _FileCache.FileCache;
exports.FileVersionNotifier = _FileVersionNotifier.FileVersionNotifier; /**
                                                                         * Copyright (c) 2017-present, Facebook, Inc.
                                                                         * All rights reserved.
                                                                         *
                                                                         * This source code is licensed under the BSD-style license found in the
                                                                         * LICENSE file in the root directory of this source tree. An additional grant
                                                                         * of patent rights can be found in the PATENTS file in the same directory.
                                                                         *
                                                                         * 
                                                                         * @format
                                                                         */

const OPEN_FILES_SERVICE = exports.OPEN_FILES_SERVICE = 'OpenFilesService';

function getBufferAtVersion(fileVersion) {
  (0, _assert2.default)(fileVersion.notifier instanceof _FileCache.FileCache, "Don't call this from the Atom process");
  return fileVersion.notifier.getBufferAtVersion(fileVersion);
}
